### PR TITLE
feature_method_to_disable_builtin_function

### LINKED ISSUE #112: disabling built-in functions
Is there a way to disable the [built-in functions](https://docs.rs/evalexpr/8.1.0/evalexpr/index.html#builtin-functions) other than simply replacing the function names in input `String`s?

### PR DESCRIPTION
**Pull request by hexofyore**
# 112- disabling built-in functions

Builtin functions can be disabled in context.

## Description

In the context, Builtin functions can be disabled. If builtin functions are called, it returns Error. 

List here your changes

- I added the method to disable builtin functions.

## Type of change

- [x] New feature (non-breaking change which adds functionality)
- [x] This change requires a documentation update

## Checklist

- [x] My code follows the contribution guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] My changes generate no new warnings
- [x] I formatted the code with `cargo fmt`
- [x] I checked my code using `cargo clippy` and reports no warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] The changes I've made are Windows, Linux compatible

Closes #112

### CODE REVIEW

**Comment by hexofyore** [comment]
@ISibboI

File: src/context/mod.rs [hunk 1]
```diff
@@ -79,6 +85,12 @@ impl Context for EmptyContext {
             identifier.to_string(),
         ))
     }
+    /// Builtin functions can't be disbaled for Empty Context.
+    fn disable_builtin_fn(&mut self) {}
+    /// Builtin functions are always disabled for Empty Context.
+    fn is_builtin_fn_disabled(&self) -> bool {
+        true
```
Comment (ISibboI) [inline src/context/mod.rs:92]: It may be useful to have a separate new type `EmptyContextWithBuiltinFunctions` that behaves like the empty context but uses builtin functions by default, and builtin functions cannot be disabled.

**Comment by hexofyore** [comment]
What happened with doc? I reverted change on readme and changed lib.rs file.

File: src/context/mod.rs [hunk 2]
```diff
@@ -105,6 +117,7 @@ pub struct HashMapContext {
     variables: HashMap<String, Value>,
     #[cfg_attr(feature = "serde_support", serde(skip))]
     functions: HashMap<String, Function>,
+    without_builtin_fn: bool,
```
Comment (ISibboI) [inline src/context/mod.rs:120]: Should be `without_builtin_functions`.

File: src/lib.rs [hunk 3]
```diff
@@ -270,6 +272,12 @@
 //!            Err(EvalexprError::ExpectedInt { actual: Value::from(5.5) }));
 //! // Reading a variable does not require a mutable context
 //! assert_eq!(eval_with_context("a", &context), Ok(Value::from(5)));
+//! // Builtin functions are enabled by default.
```
Comment (ISibboI) [inline src/lib.rs:275]: Better have this part in the `### Builtin Functions` section.

File: src/lib.rs [hunk 4]
```diff
@@ -377,6 +385,9 @@
 //! Otherwise, a float is returned.
 //!
 //! The regex functions require the feature flag `regex_support`.
+//!
```
Comment (ISibboI) [inline src/lib.rs:388]: Better talk about this at the beginning of the `### Builtin functions` section.

File: tests/integration.rs [hunk 5]
```diff
@@ -2243,3 +2243,14 @@ fn test_negative_power() {
...
+#[test]
+fn test_disabling_builtin_fn() {
+    let mut context = HashMapContext::new();
+    // Built in functions are enabled by default.
+    assert_eq!(eval_with_context("max(1,3)",&context),Ok(Value::from(3)));
+    // Disabling builtin function in Context.
+    context.disable_builtin_fn();
+    // Builting functions are disabled and using them returns Error.
+    assert_eq!(eval_with_context("max(1,3)",&context),Err(EvalexprError::FunctionIdentifierNotFound(String::from("max"))));
+}
```
Comment (ISibboI) [inline tests/integration.rs:2256]: * [ ] Should also add some test case to verify that the empty context has builtin functions disabled by default, and if you decide to add an `EmptyContextWithBuiltinFunctions`, then a test case that they are enabled.
 * [ ] Also, test the errors reported when trying to enable/disable builtin functions for the empty contexts.

**Review by ISibboI — CHANGES_REQUESTED** [review]
Thank you a lot for the pull request! It is already in a pretty nice state, but there are some comments that we should still address.

**Comment by ISibboI** [comment]
> What happened with doc? I reverted change on readme and changed lib.rs file.

You need to execute [sync-readme](https://crates.io/crates/cargo-sync-readme)

**Comment by hexofyore** [comment]
Ok. I will change them.

File: src/context/mod.rs [hunk 6]
```diff
@@ -22,6 +22,12 @@ pub trait Context {
     /// Calls the function that is linked to the given identifier with the given argument.
     /// If no function with the given identifier is found, this method returns `EvalexprError::FunctionIdentifierNotFound`.
     fn call_function(&self, identifier: &str, argument: &Value) -> EvalexprResult<Value>;
+
+    /// Checks if builtin function has been disabled.
+    fn is_builtin_fn_disabled(&self) -> bool;
+
+    /// Disables Builtin function.
+    fn disable_builtin_fn(&mut self);
```
[the remark this answers is not in the document]
Comment (hexofyore) [reply src/context/mod.rs:30]: I made it so that EmptyContext can't have builtin functions enabled while EmptyContextWithBuiltinFunctions can't have builtin functions disabled. 
They return EvalexprError::InvalidContext error when calling the setter function.

File: src/context/mod.rs [hunk 7]
```diff
@@ -22,6 +22,12 @@ pub trait Context {
     /// Calls the function that is linked to the given identifier with the given argument.
     /// If no function with the given identifier is found, this method returns `EvalexprError::FunctionIdentifierNotFound`.
     fn call_function(&self, identifier: &str, argument: &Value) -> EvalexprResult<Value>;
+
+    /// Checks if builtin function has been disabled.
+    fn is_builtin_fn_disabled(&self) -> bool;
```
[the remark this answers is not in the document]
Comment (hexofyore) [reply src/context/mod.rs:27]: renamed to are_builtin_functions_disabled

**Comment by ISibboI** [comment]
Thanks, I can take it from here.

**Comment by ISibboI** [comment]
Well I have integrated the changes now, but it seems that github does not recognise that. Hence I will have to close this, even though it is technically merged. Thank you a lot for the contribution!
